core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/kubernetesModels.scala

```diff
@@ -133,13 +155,137 @@ final case class Nodepool(id: NodepoolLeoId,
-                          errors: List[RuntimeError])
+                          autoscalingEnabled: Boolean,
+                          autoscalingConfig: Option[AutoscalingConfig],
+                          errors: List[KubernetesError],
+                          apps: List[App])
+object KubernetesNameUtils {
+  //UUID almost works for this use case, but kubernetes names must start with a-z
+  def getUniqueName[A](apply: String => A): Either[Throwable, A] =
+    KubernetesName.withValidation('k' + UUID.randomUUID().toString.toLowerCase.substring(1), apply)
+}
 
-final case class NodepoolAutoscaling(autoScalingMin: AutoScalingMin, autoScalingMax: AutoScalingMax)
+final case class AutoscalingConfig(autoscalingMin: AutoscalingMin, autoscalingMax: AutoscalingMax)
 
 final case class NodepoolLeoId(id: Long) extends AnyVal
 final case class NumNodes(amount: Int) extends AnyVal
-final case class AutoScalingMin(amount: Int) extends AnyVal
-final case class AutoScalingMax(amount: Int) extends AnyVal
+final case class AutoscalingMin(amount: Int) extends AnyVal
+final case class AutoscalingMax(amount: Int) extends AnyVal
+
+final case class DefaultKubernetesLabels(googleProject: GoogleProject,
+                                         appName: AppName,
+                                         creator: WorkbenchEmail,
+                                         serviceAccount: WorkbenchEmail) {
+  val toMap: LabelMap =
+    Map(
+      "appName" -> appName.toString,
+      "googleProject" -> googleProject.toString,
+      "creator" -> creator.toString,
+      "clusterServiceAccount" -> serviceAccount.toString
+    )
+}
+
+final case class KubernetesError(errorMessage: String, errorCode: Int, timestamp: Instant, errorSource: ErrorSource)
+sealed abstract class ErrorSource
+object ErrorSource {
+  case object Cluster extends ErrorSource {
+    override def toString: String = "Cluster"
+  }
+
+  case object Nodepool extends ErrorSource {
+    override def toString: String = "Nodepool"
+  }
+
+  case object App extends ErrorSource {
+    override def toString: String = "App"
+  }
+
+  def values: Set[ErrorSource] = sealerate.values[ErrorSource]
+  def stringToObject: Map[String, ErrorSource] = values.map(v => v.toString -> v).toMap
+}
+
+sealed abstract class AppType
+object AppType {
+  case object Galaxy extends AppType {
+    override def toString: String = "GALAXY"
+  }
+
+  def values: Set[AppType] = sealerate.values[AppType]
+  def stringToObject: Map[String, AppType] = values.map(v => v.toString -> v).toMap
+}
+
+final case class ReleaseName(value: String) extends AnyVal
+
+final case class AppId(id: Long) extends AnyVal
+final case class AppName(value: String) extends AnyVal
+//These are async from the perspective of Front Leo saving the app record, but both must exist before the helm command is executed
+final case class AppResources(namespace: Namespace, disk: Option[PersistentDisk], services: List[KubernetesService])
```

Comment:
I held off on kubernetes service account handling since the design wasn't finalized when I was coding this up. 

Not sure exactly what you mean by the namespace modeling, this ADT does have a field for the app namespace if that's what you are referring to?

Reply:
> I held off on kubernetes service account handling since the design wasn't finalized when I was coding this up.

Cool. It'd be good to capture that somewhere (ideally on an existing or new ticket?).

Reply:
I have a placeholder for it in this ticket https://broadworkbench.atlassian.net/browse/IA-1911